Add warnings blog post

liggitt opened the pull request:
cc @mkorbi @RCantw3ll

mkorbi commented:
thank you jordan!
/hold till release date 03.09.

mkorbi commented on content/en/blog/_posts/2020-09-03-warnings/index.md:
```diff
@@ -0,0 +1,272 @@
+in each release can be daunting, even for people who work on the project full-time. One important type of change is API deprecations.
+As APIs in Kubernetes graduate to GA versions, pre-release API versions are deprecated and eventually removed.
+
+Even though there is an [extended deprecation period](https://kubernetes.io/docs/reference/using-api/deprecation-policy/),
```
better only /docs/reference/using-api/deprecation-policy/

mkorbi commented on content/en/blog/_posts/2020-09-03-warnings/index.md:
```diff
@@ -0,0 +1,272 @@
+# Custom Resource Definitions
+
+Along with the API server ability to warn about deprecated API use, starting in v1.19, a CustomResourceDefinition can indicate a 
+[particular version of the resource it defines is deprecated](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-deprecation).
```
same here can do it direct as /docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-deprecation

mkorbi commented on content/en/blog/_posts/2020-09-03-warnings/index.md:
```diff
@@ -0,0 +1,272 @@
+
+# Admission Webhooks
+
+[Admission webhooks](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers)
```
and another one /docs/reference/access-authn-authz/extensible-admission-controllers

mkorbi reviewed (commented):
some small adjustments for the link

liggitt commented:
made all docs links relative

castrojo commented on content/en/blog/_posts/2020-09-03-warnings/index.md:
```diff
@@ -0,0 +1,272 @@
+in each release can be daunting, even for people who work on the project full-time. One important type of change is API deprecations.
+As APIs in Kubernetes graduate to GA versions, pre-release API versions are deprecated and eventually removed.
+
+Even though there is an [extended deprecation period](/docs/reference/using-api/deprecation-policy/),
```
This deprecation policy url is 404ing for me. Sends me to `https://github.com/kubernetes/website/blob/1ff24eb921850f62180334cd210ec7d3bc3e9ab0/docs/reference/using-api/deprecation-policy`?

liggitt replied:
links from the github preview aren't accurate

from https://deploy-preview-23592--kubernetes-io-master-staging.netlify.app/blog/2020/09/03/warnings/ it links to https://deploy-preview-23592--kubernetes-io-master-staging.netlify.app/docs/reference/using-api/deprecation-policy/ and works

castrojo replied:
Thank you, finally figured it out!

castrojo commented:
/lgtm

The indepth technical content is above my head so approver please take that into account, everything else lgtm!

mrbobbytables commented:
/approve
Thanks!

sttts commented on content/en/blog/_posts/2020-09-03-warnings/index.md:
```diff
@@ -0,0 +1,272 @@
+and a `removed_version` label that indicates the Kubernetes release in which the API will no longer be served.
+
+This is an example query using `kubectl`, [prom2json](https://github.com/prometheus/prom2json),
+and [jq](https://stedolan.github.io/jq/) to determine which deprecated APIs are being requested:
```
mention that this metrics info is temporary for that API server binary.

liggitt replied:
done

sttts commented on content/en/blog/_posts/2020-09-03-warnings/index.md:
```diff
@@ -0,0 +1,272 @@
+
+We also realized that the person *using* a deprecated API is often not the same person responsible for upgrading the cluster,
+so we added two administrator-facing tools to help track use of deprecated APIs and determine when upgrades are safe.
+
```
I would expect to see the header name here for non-client-go clients.

liggitt replied:
called that out and linked to the RFC in the intro

liggitt commented:
thanks for the technical review, addressed comments in latest commit, diff at https://github.com/kubernetes/website/compare/1ff24eb921850f62180334cd210ec7d3bc3e9ab0..67fb64e75f2acdf49a9d69e6eee5b74de58f718e?short_path=27f0ade#diff-27f0ade45010b7bbae687315f540fb9c

castrojo commented:
/lgtm

celestehorgan commented on content/en/blog/_posts/2020-09-03-warnings/index.md:
```diff
@@ -0,0 +1,274 @@
+---
+layout: blog
+title: "Warning: Helpful Warnings Ahead"
```
No review, just ❤️

liggitt commented:
updated heading levels to fit with h1 title, and to make the table of contents appear on the right side

castrojo commented:
/lgtm

sftim commented on content/en/blog/_posts/2020-09-03-warnings/index.md:
```diff
@@ -0,0 +1,274 @@
+Because the warning originates at the server, and is intercepted at the client level, it works for all kubectl commands,
+including high-level commands like `kubectl apply`, and low-level commands like `kubectl get --raw`:
+
+![kubectl applying a manifest file, then displaying a warning message 'networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress'.](kubectl-warnings.png)
```
nit: on a high-DPI screen, this screenshot looks blurry.

You could also use a terminal with fewer rows to save a bit of page real estate.

Don't let this feedback block a merge!

liggitt replied:
weird... I intentionally took the screenshots on a high-DPI screen... not sure how to improve that

I'll crop the empty lines at the bottom

sftim replied:
The blog stylesheet expands images to be the width of the page, so maybe this is a styling thing. If we never rendered the terminal image more than 80em wide for example, the pixels from the source image wouldn't render so large.

Something for another day, for sure.

liggitt replied:
I think it's some netlify or css scaling... when I look at the images locally, they look sharp

sftim commented on content/en/blog/_posts/2020-09-03-warnings/index.md:
```diff
@@ -0,0 +1,274 @@
+`kubectl` added a `--warnings-as-errors` option in v1.19. When invoked with this option,
+`kubectl` treats any warnings it receives from the server as errors and exits with a non-zero exit code:
+
+![kubectl applying a manifest file with a --warnings-as-errors flag, displaying a warning message and exiting with a non-zero exit code.](kubectl-warnings-as-errors.png)
```
nit: on a high-DPI screen, this screenshot looks blurry.

Love the alt attribute!

sftim commented on content/en/blog/_posts/2020-09-03-warnings/index.md:
```diff
@@ -0,0 +1,274 @@
+we're already considering other ways we can use this to improve people's experience with Kubernetes.
+A couple areas we're looking at next are warning about [known problematic values](http://issue.k8s.io/64841#issuecomment-395141013)
+we cannot reject outright for compatibility reasons, and warning about use of deprecated fields or field values
+(like selectors using beta os/arch node labels, [deprecated in v1.14](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.14.md#deprecations)).
```
You can if you want to link specifically to https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/#beta-kubernetes-io-os-deprecated and https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/#beta-kubernetes-io-arch-deprecated

liggitt replied:
updated

sftim commented on content/en/blog/_posts/2020-09-03-warnings/index.md:
```diff
@@ -0,0 +1,274 @@
+
+In Kubernetes v1.19, we added a feature that allows the Kubernetes API server to
+[send warnings to API clients](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1693-warnings).
+The warning is sent using a [standard `Warning` response header](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.46),
```
```suggestion
The warning is sent using a [standard `Warning` response header](https://tools.ietf.org/html/rfc7234#section-5.5),
```

(RFC 2616 has been superseded)

liggitt replied:
updated

sftim commented on content/en/blog/_posts/2020-09-03-warnings/index.md:
```diff
@@ -0,0 +1,274 @@
+
+The output shows that only read requests are being made to these APIs, and the most requests have been made to watch the deprecated Ingress API.
+
+This information could also be obtained via a Prometheus query.
```
nit; how about:
```suggestion
You can also find that information through a Prometheus query,
```

liggitt replied:
done

sftim reviewed (commented):
One thing I'd fix (link to a more recent RFC to describe the HTTP header `Warning`); along with some other optional feedback.

sftim commented:
BTW I think if it's easier, it's OK to publish this as-is and then immediately PR a fix to the RFC link.

liggitt commented:
comments addressed in https://github.com/kubernetes/website/compare/ea3827438184e7c5ff7f2aaf538b2547d13bfe0f..e3e63a0a4de181fef45bc47bd918070a660fc789

liggitt commented:
fixed the blurry terminal screenshot issue on big screens by indicating the preferred width of the images, while still constraining them to be no wider than 100% on small screens

sftim commented:
LGTM based on https://github.com/kubernetes/website/pull/23592#issuecomment-685777775 plus changes since then.

/lgtm

liggitt commented:
/hold cancel

Release date!
